applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -1025,6 +1046,170 @@ static void send_enqueued_report(struct subscriber *sub)
 	}
 }
 
+static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info *rep, uint8_t err);
+
+static int send_hid_out_report(struct bt_hogp *hogp, const uint8_t *data, size_t size)
+{
+	uint8_t report_id = data[0];
+	struct bt_hogp_rep_info *out_rep = bt_hogp_rep_find(hogp,
+							    BT_HIDS_REPORT_TYPE_OUTPUT,
+							    report_id);
+
+	if (!out_rep) {
+		/* HID output report with given ID is not supported by the peripheral. */
+		return 0;
+	}
+
+	int err = bt_hogp_rep_write_wo_rsp(hogp, out_rep,
+					   &data[1], size - 1,
+					   hogp_out_rep_write_cb);
+
+	if (err == -ENOTCONN) {
+		LOG_INF("Cannot forward, peer disconnected");
+		err = 0;
+	}
+
+	return err;
+}
+
+static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info *rep, uint8_t err)
+{
+	/* Make sure that the handler will not preempt system workqueue
+	 * and it will not be preempted by system workqueue.
+	 */
+	__ASSERT_NO_MSG(!k_is_in_isr());
+	__ASSERT_NO_MSG(!k_is_preempt_thread());
+
+	if (err) {
+		LOG_WRN("Failed to forward HID out report (err %" PRIu8 ")", err);
+	}
+
+	struct hids_peripheral *per = CONTAINER_OF(hogp, struct hids_peripheral, hogp);
+
+	/* Send update if it was queued. */
+	uint8_t sent_report_id = bt_hogp_rep_id(rep);
+	sys_snode_t *t;
+	struct enqueued_out_report *item = NULL;
+
+	SYS_SLIST_FOR_EACH_NODE(&per->enqueued_out_reports, t) {
+		struct enqueued_out_report *i = CONTAINER_OF(t, struct enqueued_out_report, node);
+
+		uint8_t i_report_id = i->data[0];
+
+		if (i_report_id == sent_report_id) {
+			item = i;
+			break;
+		}
+	}
+
+	if (!item) {
+		return;
+	}
+
+	int send_err = send_hid_out_report(hogp, item->data, item->data_size);
+
+	if (send_err) {
+		LOG_ERR("Cannot forward HID report (err: %d)", send_err);
+	}
+
+	sys_slist_find_and_remove(&per->enqueued_out_reports, &item->node);
+	k_free(item->data);
+	k_free(item);
+}
+
+static void append_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+{
+	struct enqueued_out_report *item = k_malloc(sizeof(*item));
+
+	if (!item) {
+		LOG_ERR("Failed to allocate memory to forward HID output report");
+		return;
+	}
+
+	item->data_size = size;
+	item->data = k_malloc(size);
+
+	if (!data) {
+		k_free(item);
+		LOG_ERR("Failed to allocate memory to forward HID output report");
+		return;
+	}
+
+	memcpy(item->data, data, size);
+	sys_slist_append(list, &item->node);
+};
+
+static void enqueue_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+{
+	uint8_t report_id = data[0];
+	bool updated = false;
+
+	/* If report is already present, update the data associated with report. */
+	sys_snode_t *t;
```

Comment:
Please name the variables more significantly. Especially `i`, `t`

Reply:
done